### PR TITLE
Light friendly storage tracking: track last-modification-time entry in storage

### PR DESCRIPTION
**Pull request by svyatonik**
This PR implements "Track last-modification-time entry in storage" section of #131 .

### Brief overview:
1) runtime decides if it uses prefixes in the storage
2) runtime saves 2 values in the storage when it is going to use prefixed values - ":prefix_len" and ":prefix"
2.1) all storage values are supposed to have prefix of ":prefix_len" (if the entry ":prefix_len" exists)
2.2) all new storage values are prepended by prefix from ":prefix" entry
2.3) both values are inserted at the genesis (see TODO below)
2.4) ":prefix" is updated (with encoded block#) when new blocks starts execution
3) prefixes are appended/stripped by `runtime-io` functions
4) when runtime deletes storage value, it is not actually deleted, but:
4.1) prefix is written to the storage entry
4.2) deleted key is inserted into special `KeysSet` structure (with ":deleted" prefix)
4.3) when new blocks starts execution, runtime checks if it is a good time to purge deleted values (if current_block_number % purge_interval == 0)
4.4) purge traverses `KeysSet` contents and for each deleted key checks (before purging this key):
4.4.1) if it is still deleted
4.4.2) if current_block_number - deleted_at_block_number > min_age_before_purge
4.5) storage purge is not a part of consensus (i.e. there's no guarantee that old values will be purged every N blocks)
5) there are now several places where substrate should know if values are prefixed:
5.1) when it reads values directly from the storage (e.g. ":code")
5.2) when it abouts to delete keys with given prefix
6) because of (5), there are 2 pieces of code which are shared by substrate && runtime:
6.6) to strip prefix, file `substrate/runtime-io/src/prefix_shared.rs` is compiled both by `runtime-io` and `substrate-state-machine`
6.7) to remember keys that we need to purge (when deleting by prefix), file `substrate/runtime-io/src/keys_set.rs` is compiled both by `runtime-io` and `substrate-state-machine`

### TODOs (in order of importance)
~~1) ":prefix" or ":prefix_len" are read on every storage access by `runtime-io`. We need to cache these values at least for whole "CodeExecutor::call" duration. This could be done if `Ext` would provide some kind of HashMap<&'static [u8], Vec<u8>>-based cache~~ done
~~1.5) right now high-level tests for this new functionality are mostly located in `demo-executor` crate. Would be good to have low-level tests in `substrate-runtime-system` crate. +also add the test for temporary values deletion (make sure they're not scheduled for the purge) - right now this is a transparent (hardcoded in state-root) part of high-level tests;~~
2) add support (at substrate level) for runtimes to migrate storage from non-prefixed to prefixed and back forth (this would be slow since it would affect all values in the storage)
3) provide value(s) change logs to light nodes && use this to remove mined transactions/extrinsics from the pool
4) add support (at substrate level) for runtimes to change prefix length

### TODOs for potential issues
~~1) I've found that [here](https://github.com/paritytech/polkadot/blob/master/substrate/state-machine/src/ext.rs#L124) we only remove values from the backend, while leaving values in the overlay. Is this intended (@pepyakin ?)? Used the same strategy in `save_pefix_keys`, but imho it is a potential issue. UPD: https://github.com/paritytech/polkadot/issues/426~~ fixed

### CODE REVIEW

**Comment by gavofyork** [comment]
Hmm, my preference is for the "Ordered, indexed, Merklised per-block change-trie" idea rather than the prefixing idea, but happy to play in this direction to see how it goes. One particularly problematic case is there we have a huge referendum proposal (last one was 700KB) and its existence needs to be tested (this happens several times every block). In the current model, we only need to test existence of the key in the Merkle tree; if it's there we don't actually need to load the value to know that the proposal "exists". In this new model that's not enough since it might be there but in a zombie "deleted" state - we'll have to load the value (at least the first 4 bytes) to check whether it really is in the trie. That may be problematic to do efficiently without incurring the full cost of the 700KB load.

> storage purge is not a part of consensus (i.e. there's no guarantee that old values will be purged every N blocks)

It must be since it alters the storage trie root.

**Comment by svyatonik** [comment]
@gavofyork 
1) thanks for sharing info about large values - this must be optimized as you've told (to read prefix only). Will do that in parallel with TODO#1. Right now this should't be a problem, because this PR doesn't alters any existing chain (they have no ":prefix_len" entry) && if there are no prefix, `exists_storage` works as before
2) storage root check is a part of consensus, of course. And given that the storage is the same on all nodes && ":code" runs deterministically, all scheduled values  will be purged (as it is in the current implementation). I'm talking here about potential 2nd level consensus rule - like "ensure that there are no footprints after block N%interval==0". This is redundant. Probably ignore this statement at all :)

File: substrate/executor/src/wasm_executor.rs [hunk 1]
```diff
@@ -227,6 +227,18 @@ impl_function_executor!(this: FunctionExecutor<'e, E>,
 		this.ext.clear_prefix(&prefix);
 		Ok(())
 	},
+	ext_set_prefix(prefix_data: *const u8, prefix_len: u32, value_data: *const u8, value_len: u32) => {
+		let prefix = this.memory.get(prefix_data, prefix_len as usize).map_err(|_| UserError("Invalid attempt to determine prefix in ext_clear_prefix"))?;
+		let value = this.memory.get(value_data, value_len as usize).map_err(|_| UserError("Invalid attempt to determine value in ext_clear_prefix"))?;
+		this.ext.set_prefix(&prefix, value);
+		Ok(())
+	},
+	ext_save_pefix_keys(prefix_data: *const u8, prefix_len: u32, set_prefix_data: *const u8, set_prefix_len: u32) => {
```
Comment (snd) [inline substrate/executor/src/wasm_executor.rs]: i think there's an `r` missing: `ext_save_prefix_keys`

**Comment by gavofyork** [comment]
it would be good to know if it's even possible to read the first few bytes of a storage item without incurring the cost of reading all from disk, since testing for existence is a pretty important thing to be able to do cheaply.

**Comment by svyatonik** [comment]
Found some issues when tried to run local testnet with prefixes => gotissues for now.

**Comment by svyatonik** [comment]
Added `cached_storage()` method to `Ext`. Caching all storage values seems a bad solution => only ":prefix" and ":prefix_len" are cached for now. This had slightly improved performance. Times are:
- prefix-enabled runtime + cache is disabled: `0.2910150000000001`;
- prefix-enabled runtime + cache is enabled: `0.27259384615384613`;
- no-prefix runtime + cache is enabled: `0.2519619047619048`;
- no-prefix runtime + master branch (not prefix manipulations at all): `0.24247500000000002`.

**Comment by gavofyork** [comment]
Superceded by #628 . Reopen later if that doesn't work out (though I expect it will).

**Pull request closed without merging**